### DISCUSSION TITLE
Reactive Routes vs RestEasy Reactive

### DISCUSSION THREAD

**Discussion by justinjohnson02** [quarkusio/quarkus#25963] [Q&A] [answered]
Question - is there a difference in the future roadmap/direction for Quarkus between Reactive Routes and RestEasy Reactive?

While I have experience with Vert.x, I don't have as much with Mutiny/RestEasy Reactive and enjoy the flexibility that comes with `RoutingContext` - though I am more than open to seeing how to "migrate" to Mutiny.

In the end though, I just wanted to try and "align" our code somewhat with the direction of Quarkus when it comes to REST extension development.

Thanks and awesome work to the entire team + contributors

[a comment by an automated account is not shown]

**Reply by geoand** [reply]
RESTEasy Reactive is the stack we are focusing on going forward

**Reply by justinjohnson02** [reply]
Ok thanks!

**Reply by ghubstan** [reply]
All my db repository implementations return a `io.vertx.core.Future` (fully async, "i/o" wise).

I thought this would work well with a REST api based on Reactive Routes.  I also thought my endpoint methods could return `Uni`s.  I like the Mutiny api's support for retry and failover conditions.

Now I see this discussion, so I need to base my REST api on RESTEasy Reactive, not Reactive Routes.

What would an "async" RESTEasy Reactive endpoint look like when the return value comes from a `io.vertx.core.Future`?

For example, how can this endpoint return a `Uni<String>` created from a Vertx Future?

```

    @GET
    public Uni<String> seedVendor(HttpServerRequest request, HttpServerResponse response) {
        var vendorId = 1_000_000L;

        Promise<VendorDocument> promisedDocument = Promise.promise();
        
        gatewayService.getVendorDocument(vendorId).onComplete(ar -> {
            if (ar.succeeded()) {
                promisedDocument.complete(ar.result()); 
                // Is my best option to write the value directly into the HttpServerResponse
                // now (here),  and change the method signature to return Uni<Void> ?
            } else {
                promisedDocument.fail(ar.cause());      // Cannot return error msg from here.
            }
        });

        return Uni.createFrom().item(() -> "How to do I return the promisedDocument.future().onComplete's value?");
    }

```

**Reply by ghubstan** [reply]
Or I have to use Mutiny Vert.x bindings of the Vert.x Postgres client, and make all my repo methods return Uni instead of vertx Future?

**Reply by geoand** [reply]
You can use the [Smallrye Mutiny Vert.x bindings ](https://smallrye.io/smallrye-mutiny-vertx-bindings/latest/) to convert between Vert.x and Mutiny types

**Reply by ghubstan** [reply]
Can I leave my repo classes alone, keep them returning `io.vertx.core.Future`s, and use the bindings in the calling (web-service) layer of the app?   I see my Quarkus app already imports `smallrye-mutiny-vertx-core`, and related binding modules.

I just don't see how I can transform a `io.vertx.core.Future`s completion result into a 'Uni<?>', using the bindings?

**Reply by geoand** [reply]
> Can I leave my repo classes alone, keep them returning io.vertx.core.Futures, and use the bindings in the calling (web-service) layer of the app

I don't see why not. 

> I just don't see how I can transform a io.vertx.core.Futures completion result into a 'Uni<?>', using the bindings?

There is `io.smallrye.mutiny.vertx.UniHelper#toUni`

**Reply by ghubstan** [reply]
Thanks.  I just found the helpers in `smallrye-mutiny-vertx-runtime`.  Once again, I embarrassed myself by posting a question, then finding out myself a minute later ;-/

**Reply by ghubstan** [reply]
Perfect, in the web gateway service layer, I can make calls to my _vertx-only_ db repos:

```
io.vertx.core.Future<VendorRecord> vertxFuture = vendorRepository.fetchVendorWithId(id);
Uni<VendorRecord> uni = UniHelper.toUni(vertxFuture);
```

I need to understand the 'async' implications of `Uni.createFrom().completionStage(future.toCompletionStage())`, but I know where to get the answers now.  Thanks for your patience.

**Reply by geoand** [reply]
🙏🏼
